spec/models/ability_spec.rb

```diff
@@ -89,11 +89,104 @@
       let(:other_self_organized_track) { create(:track, :self_organized) }
 
       it{ should be_able_to(:manage, user) }
-
       it{ should be_able_to(:manage, registration_public) }
       it{ should be_able_to(:manage, registration_not_public) }
-      it{ should_not be_able_to(:new, Registration.new(conference_id: conference_with_closed_registration.id))}
-      it{ should_not be_able_to(:create, Registration.new(conference_id: conference_with_closed_registration.id))}
+
+      # Test for user can register or not
+      context 'when user is not a speaker with event confirmed' do
+        let(:conference) { create(:conference) }
+        let(:registration_period) { create(:registration_period) }
+
+        context 'when the registration is closed' do
+          let(:conference_with_closed_registration) { create(:conference) }
+
+          it{ should_not be_able_to(:new, Registration.new(conference: conference_with_closed_registration)) }
+          it{ should_not be_able_to(:create, Registration.new(conference: conference_with_closed_registration)) }
+        end
+
+        context 'when the registration period is not set' do
+          it{ should_not be_able_to(:new, Registration.new(conference: conference)) }
+          it{ should_not be_able_to(:create, Registration.new(conference: conference)) }
+        end
+
+        context 'when user has already registered' do
+          let(:user_registration) { create(:registration, conference: conference, user: user) }
+
+          it{ should_not be_able_to(:new, Registration.new(conference_id: user_registration.conference_id)) }
+          it{ should_not be_able_to(:create, Registration.new(conference_id: user_registration.conference_id)) }
+        end
+
+        context 'when user has not registered and registrations are open with no registration_limit_exceeded' do
+          let(:conference_with_open_registration_and_limit_not_excedeed) { create(:conference, registration_limit: '1', registration_period: registration_period) }
```

Comment:
This way of associating the newly created conference with the registration_period is not very good, as it is not very explicit which conference the registration_period should belong to.

Why do you need to create a new conference here? Isn't it enough that we have a conference (or should have) with a registration period set? Then you can manipulate the dates of that registration_period and register your user in that conference.

Reply:
Yes we can do that! The only thing that i was trying to do is that it would make easier to understand the code!

Reply:
One of the main thing about let statement is that the value will be cached
across multiple calls in the same example but not across examples. That's why i don't think it's a problem. Please correct me if i am wrong!